File: web/src/shared/types/insights.ts
export type Confidence = 'strong' | 'weak' | 'insufficient_data'

export interface SlowQuery {
    query: string
    calls: number
    totalExecMs: number
    meanExecMs: number
}

export interface IndexCandidate {
    table: string
    seqScanCount: number
    idxScanCount: number
    suspectedColumns: string[]
    existingIndexCount: number
    writesPerSecond: number
    confidence: Confidence
    rationale: string
    selectivity?: number
}

export interface DuplicateIndex {
    table: string
    redundantIndex: string
    coveringIndex: string
    redundantColumns: string[]
    coveringColumns: string[]
    explanation: string
}

export interface UnusedIndex {
    table: string
    index: string
    sizeBytes: number
    indexScans: number
    explanation: string
}

export interface FunctionCost {
    function: string
    calls: number
    selfTimeMs: number
    isTrigger: boolean
    triggerTables?: string[]
    explanation: string
}

export interface PaginationWarning {
    query: string
    warning: string
}

export interface Insights {
    topQueries: SlowQuery[]
    indexCandidates: IndexCandidate[]
    duplicateIndexes: DuplicateIndex[]
    unusedIndexes: UnusedIndex[]
    functionCosts: FunctionCost[]
    trackFunctionsEnabled: boolean
    trackFunctionsSetting: string
    paginationWarnings: PaginationWarning[]
    nestedStatementsTracked: boolean
    statementsTrackSetting: string
    databaseSize: DatabaseSizeInfo
    connectionSaturation: ConnectionSaturation
    sequenceOverflowRisks: SequenceOverflowRisk[]
    invalidIndexes: InvalidIndex[]
    unvalidatedConstraints: UnvalidatedConstraint[]
    vacuumHealthWarnings: VacuumHealthWarning[]
    idleInTransactionWarnings: IdleInTransactionWarning[]
    checkpointHealth: CheckpointHealth
    replicationLagWarnings: ReplicationLagWarning[]
    physicalIOEnabled: boolean
    physicalIOHotspots: PhysicalIOHotspot[]
    preparedTransactionWarnings: PreparedTransactionWarning[]
    replicationSlotWarnings: ReplicationSlotWarning[]
    longRunningQueryWarnings: LongRunningQueryWarning[]
    unloggedTables: UnloggedTable[]
    lockWaitWarnings: LockWaitWarning[]
}

export interface TableSize {
    table: string
    totalBytes: number
}

export interface DatabaseSizeInfo {
    totalBytes: number
    largestTables: TableSize[]
}

export interface ConnectionSaturation {
    activeConnections: number
    maxConnections: number
    usagePercent: number
    warning?: string
}

export interface SequenceOverflowRisk {
    sequence: string
    currentValue: number
    maxValue: number
    usagePercent: number
    explanation: string
}

export interface InvalidIndex {
    table: string
    index: string
    explanation: string
}

export interface UnvalidatedConstraint {
    table: string
    constraint: string
    explanation: string
}

export interface VacuumHealthWarning {
    table: string
    deadTupleRatio: number
    lastAutovacuum?: string
    explanation: string
}

export interface IdleInTransactionWarning {
    pid: number
    user: string
    applicationName: string
    idleSeconds: number
    explanation: string
}

export interface CheckpointHealth {
    scheduledCheckpoints: number
    requestedCheckpoints: number
    requestedRatio: number
    warning?: string
}

export interface ReplicationLagWarning {
    applicationName: string
    clientAddr: string
    lagBytes: number
    explanation: string
}

export interface PhysicalIOHotspot {
    query: string
    calls: number
    execReads: number
    execWrites: number
    userTimeMs: number
    systemTimeMs: number
    explanation: string
}

export interface PreparedTransactionWarning {
    gid: string
    database: string
    owner: string
    ageSeconds: number
    explanation: string
}

export interface ReplicationSlotWarning {
    slotName: string
    active: boolean
    retainedBytes: number
    explanation: string
}

export interface LongRunningQueryWarning {
    pid: number
    user: string
    applicationName: string
    query: string
    runningSeconds: number
    explanation: string
}

export interface UnloggedTable {
    table: string
    explanation: string
}

// LockWaitWarning is MySQL-only for now — Postgres responses never include
// this field. See domain.LockWaitSession on the backend for why.
export interface LockWaitWarning {
    waitingPid: number
    waitingQuery: string
    blockingPid: number
    blockingQuery: string
    lockedTable: string
    waitAgeSeconds: number
    explanation: string
}

File: go/internal/domain/insights.go
package domain

// SlowQuery represents one normalized query's aggregate cost, as reported
// by pg_stat_statements — not a single execution, but the cumulative cost
// across every time this query shape has run.
type SlowQuery struct {
	Query       string  `json:"query"`
	Calls       int64   `json:"calls"`
	TotalExecMs float64 `json:"totalExecMs"`
	MeanExecMs  float64 `json:"meanExecMs"`
}

// IndexCandidate is a suggestion, never a certainty — the caller must
// present it as something to evaluate, not something to blindly apply.
// It always carries both the potential read benefit and the write-side
// cost of adding an index, since one without the other is misleading.
type IndexCandidate struct {
	Table              string     `json:"table"`
	SeqScanCount       int64      `json:"seqScanCount"`
	IdxScanCount       int64      `json:"idxScanCount"`
	SuspectedColumns   []string   `json:"suspectedColumns"`
	ExistingIndexCount int        `json:"existingIndexCount"`
	WritesPerSecond    float64    `json:"writesPerSecond"`
	Confidence         Confidence `json:"confidence"`
	Rationale          string     `json:"rationale"`
	Selectivity        *float64   `json:"selectivity,omitempty"`
}

// Insights bundles every category of suggestion returned by one snapshot
// of analysis. Always advisory — pgscope never applies anything itself.
type Insights struct {
	TopQueries                  []SlowQuery                  `json:"topQueries"`
	IndexCandidates             []IndexCandidate             `json:"indexCandidates"`
	DuplicateIndexes            []DuplicateIndex             `json:"duplicateIndexes"`
	UnusedIndexes               []UnusedIndex                `json:"unusedIndexes"`
	FunctionCosts               []FunctionCost               `json:"functionCosts"`
	TrackFunctionsEnabled       bool                         `json:"trackFunctionsEnabled"`
	TrackFunctionsSetting       string                       `json:"trackFunctionsSetting"`
	PaginationWarnings          []PaginationWarning          `json:"paginationWarnings"`
	NestedStatementsTracked     bool                         `json:"nestedStatementsTracked"`
	StatementsTrackSetting      string                       `json:"statementsTrackSetting"`
	DatabaseSize                DatabaseSizeInfo             `json:"databaseSize"`
	ConnectionSaturation        ConnectionSaturation         `json:"connectionSaturation"`
	SequenceOverflowRisks       []SequenceOverflowRisk       `json:"sequenceOverflowRisks"`
	InvalidIndexes              []InvalidIndex               `json:"invalidIndexes"`
	UnvalidatedConstraints      []UnvalidatedConstraint      `json:"unvalidatedConstraints"`
	VacuumHealthWarnings        []VacuumHealthWarning        `json:"vacuumHealthWarnings"`
	IdleInTransactionWarnings   []IdleInTransactionWarning   `json:"idleInTransactionWarnings"`
	CheckpointHealth            CheckpointHealth             `json:"checkpointHealth"`
	ReplicationLagWarnings      []ReplicationLagWarning      `json:"replicationLagWarnings"`
	PhysicalIOEnabled           bool                         `json:"physicalIOEnabled"`
	PhysicalIOHotspots          []PhysicalIOHotspot          `json:"physicalIOHotspots"`
	PreparedTransactionWarnings []PreparedTransactionWarning `json:"preparedTransactionWarnings"`
	ReplicationSlotWarnings     []ReplicationSlotWarning     `json:"replicationSlotWarnings"`
	LongRunningQueryWarnings    []LongRunningQueryWarning    `json:"longRunningQueryWarnings"`
	UnloggedTables              []UnloggedTable              `json:"unloggedTables"`
	// LockWaitWarnings is MySQL-only for now — see domain.LockWaitSession.
	LockWaitWarnings []LockWaitWarning `json:"lockWaitWarnings"`
}
